cunumeric/_ufunc/ufunc.py

```diff
@@ -212,6 +212,88 @@ def _maybe_cast_input(self, arr, to_dtype, casting):
 
         return arr.astype(to_dtype)
 
+    def _maybe_create_result(self, out, out_shape, res_dtype, casting, inputs):
+        if out is None:
+            return ndarray(shape=out_shape, dtype=res_dtype, inputs=inputs)
+        elif out.dtype != res_dtype:
+            if not np.can_cast(res_dtype, out.dtype, casting=casting):
+                raise TypeError(
+                    f"Cannot cast ufunc '{self._name}' output from "
+                    f"{res_dtype} to {out.dtype} with casting rule "
+                    f"'{casting}'"
+                )
+            return ndarray(shape=out.shape, dtype=res_dtype, inputs=inputs)
+        else:
+            return out
+
+    @staticmethod
+    def _maybe_cast_output(out, result):
+        if out is None or out is result:
+            return result
+        else:
+            out._thunk.convert(result._thunk, warn=False)
+            return out
+
+    @staticmethod
+    def _maybe_convert_to_cunumeric_ndarray(out):
+        if out is None:
+            return None
+        elif isinstance(out, ndarray):
+            return out
+        elif isinstance(out, np.ndarray):
+            return convert_to_cunumeric_ndarray(out)
+        else:
+            raise TypeError("return arrays must be of ArrayType")
+
+    def _prepare_operands(self, *args, out=None, where=True):
+        max_nargs = self.nin + self.nout
+        if len(args) < self.nin or len(args) > max_nargs:
+            raise TypeError(
+                f"{self._name}() takes from {self.nin} to {max_nargs} "
+                f"positional arguments but {len(args)} were given"
+            )
+
+        inputs = tuple(
+            convert_to_cunumeric_ndarray(arr) for arr in args[: self.nin]
+        )
+
+        if len(args) > self.nin:
+            if out is not None:
+                raise TypeError(
+                    "cannot specify 'out' as both a positional and keyword "
+                    "argument"
+                )
+            out = args[self.nin :]
+            # Missing outputs are treated as Nones
+            out = out + (None,) * (self.nout - len(out))
+        elif out is None:
+            out = (None,) * self.nout
+        elif not isinstance(out, tuple):
+            out = (out,)
+
+        outputs = tuple(
+            self._maybe_convert_to_cunumeric_ndarray(arr) for arr in out
+        )
+
+        if self.nout != len(outputs):
+            raise ValueError(
+                "The 'out' tuple must have exactly one entry "
+                "per ufunc output"
+            )
+
+        shapes = [arr.shape for arr in inputs]
+        shapes.extend(arr.shape for arr in outputs if arr is not None)
+
+        # Check if the broadcasting is possible
+        out_shape = broadcast_shapes(*shapes)
```

Comment:
The output array, if provided by the user, should not be broadcastable (the inputs can be broadcasted up to match it, but we shouldn't broadcast the output array itself). E.g. the following:

```
x = np.ones((4,))
y = np.ones((4,))
z = np.zeros((1,))
np.multiply(x, y, out=z)
print(z)
```

in NumPy prints:

```
ValueError: non-broadcastable output operand with shape (1,) doesn't match the broadcast shape (4,)
```

while cunumeric will crash later on:

```
  File "/Users/mpapadakis/cunumeric/cunumeric/_ufunc/ufunc.py", line 605, in __call__
    result._thunk.binary_op(self._op_code, x1._thunk, x2._thunk, where, ())
  File "/Users/mpapadakis/cunumeric/cunumeric/deferred.py", line 75, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/deferred.py", line 1744, in binary_op
    rhs1 = src1._broadcast(lhs.shape)
  File "/Users/mpapadakis/cunumeric/cunumeric/deferred.py", line 604, in _broadcast
    assert result.shape[dim] == 1
``` 

I think we can just check at this point:

```
if any(arr.shape != out_shape for arr in outputs):
  raise ValueError("...")
```

Reply:
297a5682d762acbe150023de8089f3659c3ec5da